### PR TITLE
Add user defined directory menu

### PR DESCRIPTION
**Pull request by christoph-heinrich**
~~Here is an initial implementation of the feature from https://github.com/po5/memo/pull/12#issuecomment-1680865209. I was too eager to try that out to wait for a response :smile:~~

Users can define which directories should be listed by specifying how the directory structure of it's parents should look like.

The idea is that e.g. directories inside a directory called "TV Shows" are directories of tv shows.
By defining the prefix "TV Shows" the user can list all recently watched tv shows, providing a convinient for continuing them.

~~I don't really understand why in `memo_clear()` it is necessary to set `tvshow_menu = false` before the `event_loop_exhausted` check, but otherwise it will sometimes open up the menu again after an entry has been selected.~~

Also when opening the menu it flashes a bigger menu before reducing it's size. I assume that's because I don't have enough tv shows in my history to fill up the whole page, but I've never noticed that with the regular history.

### CODE REVIEW

**Comment by po5** [comment]
This doesn't work for me, not because of any bugs but because the approach is flawed. It finds a grand total of 1 file.

> Assuming a directory structure like `<show name>/<season>/<episode>` is reasonable

It's not. Most of my files are are in flat folders like `Curb.Your.Enthusiasm.S02.NORDiC.1080p.WEB-DL.DD5.1.H.264-DTS`, and it gets a lot more fun with anime where everything is in flat folders like `[Kaleido-subs] Joshiraku (Rakugo Girls) (BD 1080p)`. That's how most releases come.

This works slightly better, and still handles Plex-style directories.

```diff
function show_history(entries, next_page, prev_page, update, return_items)
         elseif options.hide_same_dir or tvshow_menu then
             dirname, basename = mp.utils.split_path(display_path)
             if tvshow_menu then
-                if not dirname:find(".+/Season %d+/$") then
+                if not dirname:find(".+/Season %d+/$") and not dirname:find("S%d%d+") then
                     return
                 end
                 dirname, tvshow = mp.utils.split_path(dirname:sub(1, -2))
-                _, tvshow = mp.utils.split_path(dirname:sub(1, -2))
+                if tvshow:find("^Season %d+$") then
+                    _, tvshow = mp.utils.split_path(dirname:sub(1, -2))
+                end
             end
```

But it's still so unreliable that it's useless, and doesn't work with anime. Exactly the impossible problem I described in #12.  
What's wrong with a folder listing? (with some special handling for strings we know for sure are irrelevant like Season n)

```diff
@@ -711,11 +711,16 @@ function show_history(entries, next_page, prev_page, update, return_items)
         elseif options.hide_same_dir or tvshow_menu then
             dirname, basename = mp.utils.split_path(display_path)
             if tvshow_menu then
-                if not dirname:find(".+/Season %d+/$") then
-                    return
-                end
                 dirname, tvshow = mp.utils.split_path(dirname:sub(1, -2))
-                _, tvshow = mp.utils.split_path(dirname:sub(1, -2))
+                if tvshow:find("^Season %d+$") or tvshow:find("^S%d+$") then
+                    _, _tvshow = mp.utils.split_path(dirname:sub(1, -2))
+                    if _tvshow ~= "" then
+                        tvshow = _tvshow
+                    end
+                end
+                if tvshow == "" then
+                    tvshow = nil
+                end
             end
             if state.known_dirs[dirname] then
                 return
@@ -746,9 +751,9 @@ function show_history(entries, next_page, prev_page, update, return_items)
         end
 
 
-        if tvshow_menu then
+        if tvshow_menu and tvshow then
             title = tvshow
-        elseif title == "" then
+        elseif title == "" or (tvshow_menu and not tvshow) then
             if is_remote then
                 title = display_path
             else
```

It will never be perfect. Someone will name their season folder `Season 1.5` or `Part 1` and the special check will fail. Someone will have their folders named in German `Staffel 1`. Someone will be playing extras inside an `Extras` folder which we'll also need special handling for.

I love magic, when it works. Here's [a regex](https://gist.github.com/po5/e668df1650d0231d163fdbe1e8f7043f#file-ptp-group-names-user-js-L32) I wrote long ago to illustrate why I don't want to try extracting useful information from the wild west of filenames. In the same vein, I don't care if a user with `use_titles` ends up with a history full of `Episode n` because their files are tagged that way. It's as correct as I can make it.

Do you need this feature? Implementing it as a menu feels like the wrong approach, as it can't benefit from search and memo-last. Should be a toggle like `use_titles`, perhaps `use_folder_titles`. If you need it on-demand you set some script-opts before the script-binding.

-- uosc rants

The bigger menu is because we prefill empty menu items (up to the user's max entries per page +prev/next controls) while the current page is loading.  
It will only manifest itself if we don't have a full page of results after reading a page worth of entries from the history file. It's much less confusing than the menu not showing up at all for 5+ seconds after pressing the history key.  
No such padding is applied when the full page finally resolves, and when on the last page you likely won't have a full page, so the menu shrinks.  
This padding is done because we need to maintain the selected item across pages, and setting the menu to have less entries than the currently selected index will cause uosc to lose that selection. This is important for prev/next buttons.  
We could move prev/next buttons to the top of the list to avoid this issue, but then we run into the issue that we don't know if these controls will still be needed once the full page loads. It would result in the user's selection unexpectedly jumping down onto entries.  
We could also keep the padding on non-full pages, but I think it's ugly.

Your issue of the menu opening after selecting an entry is likely because reading the file is blocking, but uosc's menu is not. With a page that takes a long time to load, the user can select something in the uosc menu and we won't be notified of it until we finish loading the entire page (we'll get the `memo-clear` on-close message only once we exit the current event loop), but in the meantime we happily send `update-menu` messages, thinking it's still open. I wish uosc would ignore update-menu when there's no corresponding menu, but maybe that breaks someone else's use case?  
I want to fix that, it requires splitting the file read across event loops. The actual processing will still be blocking the event loop, it's just that we'll allow it to run once in a while, and only make menu update decisions at the very beginning of an event loop. I don't know if/by how much that will slow down results, but it will greatly improve the vanilla menu's experience, that right now gets completely locked up while processing the current page as it needs the event loop to run at all to react to keypresses.  
With this we could even fully control *when* the uosc menu gets closed (and get rid of the reopen after file select jank), by giving all items `keep_open` and proxying the `loadfile` command back through memo, we can then tell uosc to close the menu (well, to "open" it again knowing it's already displayed). This is another issue with uosc's API, there's no explicit way to close a given menu. Even with this approach if the user pressed ESC to close the menu from the uosc side, and your script happens to be sending an `open-menu` message meant to close it at that moment, the menu will end up reopening, in memo's case without any items because that's what we do in `memo_close`, there's nothing to be shown. Other than a message like "Sorry this bug is out of our control" :^)  
This is a bigger issue for scripts that can require long processing like memo, but affects everything.

**Comment by christoph-heinrich** [comment]
> Most of my files are are in flat folders

That's unexpected, literally all my shows are organized like that.

> It will never be perfect. Someone will name their season folder Season 1.5 or Part 1 and the special check will fail. Someone will have their folders named in German Staffel 1. Someone will be playing extras inside an Extras folder which we'll also need special handling for.

I didn't expect it to be perfect, but good enough for the common cases would still be useful. Supporting things like Season 1.5 Part 1 shouldn't be too hard, but supporting other languages too would explode the complexity. Users could configure their words for "Season" I guess. The Extras folder didn't even enter my mind, but not supporting it shouldn't be too big of a problem. As long as users know that only works for that specific directory structure, it wouldn't lead to any unexpected behavior.

> What's wrong with a folder listing?

The way I imagined it would end up looking like would be something like:
```
...ta/TV Shows/Curb Your Enthusiasm/Season 1/S01E01 - The Pants Tent.mkv
...hed/Nichijou - My Ordinary Life/Season 1/EP.1.v0.1639748011.1080p.mp4
...on (Mushoku Tensei: Isekai Ittara Honki Dasu)/5 Season 2/mushod_4.mp4
```
with a bunch of random things in between, which is still better then only the episode name, but much worse then a menu that only shows the show name.

> I love magic, when it works. Here's [a regex](https://gist.github.com/po5/e668df1650d0231d163fdbe1e8f7043f#file-ptp-group-names-user-js-L32) I wrote long ago to illustrate why I don't want to try extracting useful information from the wild west of filenames

Oof that regex is gnarly. How does one even come up with that? Seems almost like something auto-generated based on some examples. I'd rather not have that feature then deal with something like that.

> Do you need this feature? Implementing it as a menu feels like the wrong approach, as it can't benefit from search and memo-last.

It's very convenient, and implementing it as a menu is the simplest thing to do. The additional value of making it also work with memo-last and memo-search seems rather small to me.

> -- uosc rants
> _blocking io causing problems_

Yes handling that better would require async io. Technically that could be achieved by making a second script for dealing with io and then communicating with that via script-messages. That way it would be possible to e.g. read the file and only open a temporary menu when the loading takes more then e.g. 100ms. That would add some complexity, but would have the benefit of all the file and path handling being separate from the UI logic, and then the UI script would always be ready to react to events.

`loadfile` could already be done by memo instead of letting uosc execute the command directly, getting immediate feedback on selection. Each menu entry simply needs a unique script message command, I'm doing a similar thing [here](https://github.com/christoph-heinrich/mpv-quality-menu/blob/9bb4d87681b9765a8035158c9076f4a37b6f7b07/quality-menu.lua#L957-L961). I don't think there would be any benefit to keeping the menu open and then closing it via a command.

That race condition with closing is only a problem because of blocking io. uosc sends `on_close` once the menu is fully faded out, and if I remember that correctly, sending an `open-menu` while the menu is fading out doesn't make a new menu open up. The way it works out is
1. user closes though memo and then presses ESC
    1. memo sends `menu-open` to close the menu
    2. while the menu fades out the user presses ESC, but since the menu is already closing it doesn't do anything
    3. memo gets the message from on_close
2. users presses ESC and then closes though memo
    1. the ESC causes the menu to fade out
    2. user input causes memo to send `menu-open` to close the menu
    3. memo gets the message from on_close

Either way it's fine. Try breaking quality-menu, as it has to deal with the same kind of thing. The only thing throwing a wrench in that is opening the menu and then blocking on io. Things like that were the entire reason why I made the rewrite to call ytdl asynchronously in quality-menu.

BTW I found that representing the UI state as a sort of state machine is extremely useful. The idea was to mimic the Typestate pattern from Rust. That guided me though most of the rewrite and once it was done everything seemed to "just work" (only [two](https://github.com/christoph-heinrich/mpv-quality-menu/issues/25) [bugs](https://github.com/christoph-heinrich/mpv-quality-menu/issues/26) despite the large change).

**Comment by christoph-heinrich** [comment]
I've been thinking about how to make this more generally useful, and I think the simplest thing would be to have a config with "path prefixes". The idea being that people can define what the path looks like right before their shows, and it doesn't have to be an absolute path. E.g. if someone has all their shows in a directory called "TV Shows", then they could just define that as prefix and now when we encounter a path like "/data/TV Shows/Curb Your Enthusiasm/random.name.S1E1.mp4" we know that "Curb Your Enthusiasm" will be the show name and everything under that belongs to that show, doesn't matter if it's a flat folder or has sub-directories.
Those prefixes can also span multiple levels, so if someone were to have folders with e.g. genres, they can include those in the prefixes like "TV Shows/Comedy". That would require listing all the genres, so maybe we could allow people to define patterns? "TV Shows/.-/" or something like that.
Since that doesn't actually search for "Season %d"  or anything, it's more like a user defined folder menu, and because they don't have to be absolute paths it also works for people that have their shows on multiple drives.

**Comment by po5** [comment]
I like the path prefixes approach, smart.  
If we allow for patterns it'd be nice to have it as a separate option or require some prefix like `pattern:TV Shows/.-`. Someone wanting to truly match `pattern:` can double it up, even if `pattern:pattern:` looks a bit silly.  
An example value for the path prefixes param would be `My-Movies|pattern:TV Shows/.-/|Anime`.  
Match those and cut paths at the match end, remove leading slashes/backslashes. We then can't use `utils.split_path` because the file may be infinitely nested past our "TV" folder, so regex out the first directory ourselves.

I'm totally for this now (:

Blocking io:  
Splitting up the UI would be the most foolproof way to deal with it, but do you know of a way to have mpv load two scripts from a "script directory" independently? I doubt that's possible, we'll end up needing two lua files in the main scripts dir.

**Comment by christoph-heinrich** [comment]
> Splitting up the UI would be the most foolproof way to deal with it, but do you know of a way to have mpv load two scripts from a "script directory" independently? I doubt that's possible, we'll end up needing two lua files in the main scripts dir.

That will require two scripts in the main scripts dir, like is the case with the [mpv_thumbnail_script](https://github.com/marzzzello/mpv_thumbnail_script).

**Comment by christoph-heinrich** [comment]
The prefix thing has been implemented now. I copied the path separator and platform code from uosc.
The way it is written now the parent directory doesn't have be completely matched, so "TV Sho" would still show the directories inside "TV Shows".

While committing this I had the idea that the pattern could be supplied by a script-message instead of (or in addition to) an option.
This would enable users to bind different patterns to different keys.
```
t script-message-to memo memo-dirs "pattern:TV Shows/.-/"
a script-message-to memo memo-dirs Anime
```

File: memo.lua [hunk 1]
```diff
@@ -39,13 +39,42 @@ local options = {
...
 }
 
+function parse_path_prefixes(path_prefixes)
+    local plain = {}
+    local pattern = {}
+    for prefix in path_prefixes:gmatch('([^|]+)') do
+        if prefix:find("pattern:", 1, true) == 1 then
+            -- check for escaped pattern:
+            if prefix:find("pattern:", 9, true) == 9 then
+                plain[#plain + 1] = prefix:sub(9)
```
Comment (po5) [inline memo.lua:57]: Why is this needed? This should still act as a pattern match, what I meant is that someone *can* still inherently match folder names that start with this string, doesn't warrant special handling.

File: memo.lua [hunk 2]
```diff
@@ -39,13 +39,42 @@ local options = {
     select_binding = "RIGHT ENTER",
     append_binding = "Shift+RIGHT Shift+ENTER",
     close_binding = "LEFT ESC",
+
+    -- Path prefixes for the recent directory menu
+    -- Syntax: prefixes are separeted by | and can use Lua patterns by prefixing
```
Comment (po5) [inline memo.lua:44]: separeted -> separated, and option needs to be mirrored in memo.conf.

File: memo.lua [hunk 3]
```diff
@@ -39,13 +39,42 @@ local options = {
...
+    -- them with "pattern:", otherwise they will be treated as plain text.
+    -- Pattern syntax can be found here https://www.lua.org/manual/5.1/manual.html#5.4.1
+    path_prefixes = "My-Movies|pattern:TV Shows/.-/|Anime"
```
Comment (po5) [inline memo.lua:47]: I don't feel too good about having builtin prefixes, other than a wildcard for simple directory view. The example pattern can be in readme.

**Review by po5 — CHANGES_REQUESTED** [review]
I haven't yet been able to test the actual functionality, but needs these edits for now.

**Comment by po5** [comment]
> I have no idea how to explain that feature to users...

By adding a "Script messages" section to readme in a similar format to "Custom keybinds". Should completely replace your current keybinding.  
It may be worth converting all keybinds to script messages even if I don't think any existing keybinds can benefit from the extra flexibility script messages give us, to reduce confusion.

**Comment by christoph-heinrich** [comment]
> By adding a "Script messages" section to readme

I meant the path prefix thing in general, not the script messages.

File: memo.lua (hunk 3 above)
Comment (christoph-heinrich) [reply memo.lua:47]: That will require some special handling, after all a prefix that matches anything ends up matching the root of the filesystem...

**Comment by christoph-heinrich** [comment]
This has been working great for me...

**Comment by christoph-heinrich** [comment]
There have been a few times where I would have liked to open the directory itself and not it's most recently open file (because the file doesn't exist anymore).
Do you think it's worth adding another parameter to the script-message or a second script-message for that?

**Comment by po5** [comment]
Could make it fall back to the directory itself if the file doesn't exist but the dir does.

**Comment by christoph-heinrich** [comment]
Good idea with the fall back.
I'm not 100% sure if `effective_path` is the right one to use for this, but other then that this should be fine.

**Comment by po5** [comment]
Sorry about the lack of involvement I do want to get this merged when I set aside some time to look at it.

**Comment by christoph-heinrich** [comment]
I'm not 100% happy with that anyway.

Currently the fallback only works when checking for deleted files, which makes sense of course, but that means I have to turn on deletion checking to get the benefits of the fallback. That spins up my harddrives, which is annoyingly slow, even when I want to open something that isn't even on them.

We'd need some way to specify that deletion checking only happens on certain paths or something.
Alternatively maybe instead of directly loading the selected path, we could have a third deletion check mode where it does the check for the fallback only on selection.

That doesn't have to be part of this PR, but it's something that could be improved eventually.

**Comment by po5** [comment]
This crash happens when mp.split_path returns `.` for dirname (no dir found).  
Also for files at the root of the filesystem like `/fun.mkv` or non-nested relative paths `test/fun.mkv`

```
[memo] 
[memo] stack traceback:
[memo]  /home/eva/.config/mpv/scripts/memo.lua:802: in function 'read_line'
[memo]  /home/eva/.config/mpv/scripts/memo.lua:939: in function 'show_history'
[memo]  /home/eva/.config/mpv/scripts/memo.lua:1029: in function 'fn'
[memo]  mp.defaults:230: in function 'fn'
[memo]  mp.defaults:65: in function 'handler'
[memo]  mp.defaults:382: in function 'handler'
[memo]  mp.defaults:512: in function 'call_event_handlers'
[memo]  mp.defaults:554: in function 'dispatch_events'
[memo]  mp.defaults:505: in function <mp.defaults:504>
[memo]  [C]: in ?
[memo]  [C]: in ?
[memo] Lua error: /home/eva/.config/mpv/scripts/memo.lua:802: attempt to index local 'basename' (a nil value)
```

The first case is trivial, just add `if dirname == "." then return end` at the top of the dir_menu block.  
The other requires more changes, and I'd like to redo the logic to avoid issues like the extra assignment to dirname being absolutely wrong:
```
[memo] display_path /mnt/18t1/projects/hoard/conan/sorted/tv/English/enc/fluffy.is/Files/Anime-E/Detective Conan/Movies/Detective Conan - Movie 13 Trailer [ENG SoftSub][TheQuyen].mkv
[memo] original_basename Detective Conan - Movie 13 Trailer [ENG SoftSub][TheQuyen].mkv
[memo] original_dirname /mnt/18t1/projects/hoard/conan/sorted/tv/English/enc/fluffy.is/Files/Anime-E/Detective Conan/Movies/ 
[memo] new_basename Detective Conan
[memo] new_dirname /mnt/18t1/projects/hoard/conan/sorted/tv/English/enc/fluffy.is/Files/Anime-E/Detective Con
```

Using the example pattern. It matched up to /Anime but the directory is named Anime-E. As a result the new dirname is missing two chars.

I don't understand how the special case for UNC paths in `path_separator` applies to our use case.  
Handling Windows and Linux paths differently seems weird to me, since that makes path_prefixes with nested dirs OS-dependent. We should unconditionally replace backslashes with forward slashes for the purposes of matching, in which case I do see the need for special handling of UNC paths.

The menu title should change from "History (memo)" to "Directories (memo)".

I'll experiment some more and make the necessary changes to get it merged, posting in case you have additional insight but you don't have to make changes.

**Comment by christoph-heinrich** [comment]
I thought we're always working with absolute paths there, and that never came up because I never open files in mpv like that.

I also never expected the case where not the whole directory name gets matched, seems so obvious in retrospect. :disappointed: 

I can quickly fix the path separator, cut off directory name and menu title. Idk how you want to deal with non absolute paths.

**Comment by christoph-heinrich** [comment]
UNC paths are apparently a bit of a problem for split_path(), so if we want to support those we'll have to make our own split function.

**Comment by christoph-heinrich** [comment]
I've pushed some changes as a separate commit, so that they can be easily removed in case it conflicts with whatever you've been doing.

I haven't done any UNC handling, apparently split_path() should work fine on windows anyway, and mpv apparently never checks for UNC (except in defaults.js apparently), so I don't know if UNC paths even work outside of windows.

**Comment by christoph-heinrich** [comment]
> I know it's late into the PR to have this realization: I am also thinking about whether this should show directories, or show _stuff from directories_. In my case my main Anime directory contains both loose files and folders, whichever way they were uploaded online. Instead of showing directories within my anime folder, it would show files from directories and files that are on the same base dir level. In that case for aesthetic reasons I would want to display filenames everywhere, latest file from the dir (recursively) and regular filenames of loose files. This could make the resuming aspect more obvious, since you see which episode you were on.

I'm against showing the filename instead of the directory name, that way I'll end up with entries like `Episode x.mp4` instead of `Curb Your Enthusiasm`. That's not really helpful.

> It would end up being a better version of `hide_same_dir` that's very easy to control at runtime since it comes with the script-message. We'd get rid of `hide_same_dir` in favor of a suggested keybind that achieves the same functionality as a basic example.

How would that behave the same as `hide_same_dir` if it also shows loose files of that dir?

> Could still be made to only show directories if we have a trailing slash, like a glob.
> 
> That idea aside could the directory mode be made more versatile by using capture groups? e.g. `TV Shows/(.-)/.-/.-` looks for files such as `TV Shows/Mr. Robot/S01/S01E01.Robot.Guy.mkv`, and uses `Mr. Robot` as the display title?

That's something I've been thinking about too. Although I fail to see the use case for that specific example.

I have some cases where I have a show in the directory of another show because they belong together. E.g. `TV Shows/Game of Thrones/` contains `Season x` and also `House of the Dragon`. Currently that gets listed as "Game of Thrones", but it would be great if it was possible to specify patterns to differentiate between those somehow so that they're listed as if `House of the Dragon` is separate from the rest of `Game of Thrones`. Anything I've thought of so far didn't work out. It's not a huge problem, but maybe you have some ideas.
One thing that came from my experiments was that the patterns don't get split up into `plain` and `pattern`, and instead the original order gets preserved, which makes find_path_prefix() a bit shorter and maybe there is a use case where the order matters.

> on Linux it's either mounted

And when it's mounted the path is like any other directory path on linux? Then I guess there is no reason to have special treatment for UNC.

**Comment by po5** [comment]
> How would that behave the same as `hide_same_dir` if it also shows loose files of that dir?

That option was added as a "good enough" way to resume TV shows, but it'll stay around if the dir menu shows dir names instead of filenames.

> And when it's mounted the path is like any other directory path on linux?

Yeah. The way you did it is smarter than what I was envisioning, doesn't need anything special for UNC paths, nice.  
The path_prefixes simplification is good too.

I ran into an issue where it listed empty dirs and opening them would close the player.  
`if stat and stat.size ~= 0 then` works on Linux, can you confirm that `utils.file_info()` on an empty dir returns size 0 and on a non-empty dir returns non-zero for Windows?

Thanks for the quick replies when I was not so kind to do the same. Will merge and apply my changes, you can tell me if it doesn't work :^)  
uosc search next!

**Comment by christoph-heinrich** [comment]
> can you confirm that `utils.file_info()` on an empty dir returns size 0 and on a non-empty dir returns non-zero for Windows?

I don't use windows.

**Pull request merged**